include/nbl/system/ISystem.h

```diff
@@ -0,0 +1,244 @@
+        E_REQUEST_TYPE type;
+        std::variant<
+            SRequestParams_CREATE_READ,
+            SRequestParams_CREATE_WRITE,
+            SRequestParams_READ,
+            SRequestParams_WRITE
+        > params;
+
+        void set_skip()
+        {
+            write_lock_guard<> lk(rwlock);
+            set_skip_no_lock();
+            future->request = nullptr;
+        }
+    };
+
+    class CAsyncQueue : public ISkippableAsyncQueueDispatcher<CAsyncQueue, SRequestType, 256u>
+    {
+        using base_t = ISkippableAsyncQueueDispatcher<CAsyncQueue, SRequestType, 256u>;
+
+    public:
+        explicit CAsyncQueue(ISystem* owner, core::smart_refctd_ptr<ISystemCaller>&& caller) : m_owner(owner), m_caller(std::move(caller)) {}
+
+        template <typename RequestParams>
+        void request_impl(SRequestType& req, RequestParams&& params)
+        {
+            req.type = params.type;
+            req.params = std::move(params);
+            req.future = params.future;
+            req.retval_storage = params.retval_storage;
+        }
+
+        void process_request(SRequestType& req)
+        {
+            switch (req.type)
+            {
+            case ERT_CREATE_READ:
+            {
+                auto& p = std::get<SRequestParams_CREATE_READ>(req.params);
+                reinterpret_cast<core::smart_refctd_ptr<IReadFile>*>(req.retval_storage)[0] = m_caller->createReadFile(m_owner, p.filename);
+            }
+                break;
+            case ERT_CREATE_WRITE:
+            {
+                auto& p = std::get<SRequestParams_CREATE_WRITE>(req.params);
+                reinterpret_cast<core::smart_refctd_ptr<IWriteFile>*>(req.retval_storage)[0] = m_caller->createWriteFile(m_owner, p.filename);
+            }
+                break;
+            case ERT_READ:
+            {
+                auto& p = std::get<SRequestParams_READ>(req.params);
+                reinterpret_cast<uint32_t*>(req.retval_storage)[0] = m_caller->read(p.file, p.buffer, p.offset, p.size);
+            }
+                break;
+            case ERT_WRITE:
+            {
+                auto& p = std::get<SRequestParams_WRITE>(req.params);
+                reinterpret_cast<uint32_t*>(req.retval_storage)[0] = m_caller->write(p.file, p.buffer, p.offset, p.size);
+            }
+                break;
+            }
+        }
+
+    private:
+        ISystem* m_owner;
+        core::smart_refctd_ptr<ISystemCaller> m_caller;
+    };
+
+    CAsyncQueue m_dispatcher;
+
+
+    template <typename FutureType, typename RequestParamsType>
+    void requestAndPrepareFuture(SFuture<core::smart_refctd_ptr<IReadFile>>& future, RequestParamsType&& params)
+    {
+        auto& req = m_dispatcher.request(std::move(params));
+        future.asyncQ = &m_dispatcher;
+        future.request = &req;
+    }
+
+public:
+    explicit ISystem(core::smart_refctd_ptr<ISystemCaller>&& caller) : m_dispatcher(this, std::move(caller))
```

Comment:
huh... why do I need to create the caller myself and give it to the system when instantiating the system?

Reply:
so you can use ISystem API with whatever backend you want (just implement ISystem::ISystemCaller interface), guaranteed every system call is performed in dedicated thread

Reply:
so its the `ISystemCaller` that we implement per-platform?

In that case the `ISystem` is no longer an interface (no pure virtuals) and a concrete class I can `new/create` by itself, so it should be named `CSystem`

Reply:
> so its the ISystemCaller that we implement per-platform?

yep, i want to have all system calls in there. Actually if we're choosing to have all OS calls in dedicated thread, I'd make windows and clipboard stuff be done by ISystemCaller as well (ui::IClipboardManager should hold ref to ISystem, but not sure about design for window creation then)

Reply:
>  I'd make windows and clipboard stuff be done by ISystemCaller as well (ui::IClipboardManager should hold ref to ISystem, but not sure about design for window creation then)

When created, windows should take a reference to ISystem, I would not create them on a dedicated thread.

Whether Clipboard should or even can be done on a dedicated thread is a thing for @sadiuk to investigate